File: prototype-projects/backend/scripts/sync_typesense.py
import typesense
import psycopg2
import os

POWERNOVA_DB = os.environ['POWERNOVA_DB']
POWERNOVA_USER = os.environ['POWERNOVA_USER']
POWERNOVA_PASSWORD = os.environ['POWERNOVA_PASSWORD']
POWERNOVA_HOST = os.environ['POWERNOVA_HOST']
POWERNOVA_PORT = os.environ['POWERNOVA_PORT']
TYPESENSE_HOST=os.environ['TYPESENSE_HOST']
TYPESENSE_PORT=os.environ['TYPESENSE_PORT']
TYPESENSE_PROTOCOL=os.environ['TYPESENSE_PROTOCOL']
TYPESENSE_API_KEY=os.environ['TYPESENSE_API_KEY']

typesenseClient = typesense.Client({
    'api_key': 'xyz',
    'nodes': [{
        'host': TYPESENSE_HOST,
        'port': TYPESENSE_PORT,
        'protocol': TYPESENSE_PROTOCOL
    }],
    'connection_timeout_seconds': 2
})


def get_db_connection():
    conn = psycopg2.connect(database=POWERNOVA_DB,
                            host=POWERNOVA_HOST,
                            user=POWERNOVA_USER,
                            password=POWERNOVA_PASSWORD,
                            port=POWERNOVA_PORT)
    return conn


# Define the Typesense collection schema
schema = {
    "name": 'projects',
    "fields": [
        { "name": 'iso', "type": 'string' },
        { "name": 'queueid', "type": 'string'},
        { "name": 'county', "type": 'string' },
        { "name": 'state', "type": 'string' },
        { "name": 'gentype', "type": 'string' },
        { "name": 'description', "type": 'string' }
    ]
}

projects_schema = {
    "name": 'projects',
    "fields": [
        { 
            "name": 'iso', 
            "type": 'string' 
        }, { 
            "name": 'queueid', 
            "type": 'string'
        }, { 
            "name": 'county', 
            "type": 'string' 
        }, { 
            "name": 'state', 
            "type": 'string' 
        }, { 
            "name": 'gentype', 
            "type": 'string' 
        }, { 
            "name": 'description', 
            "type": 'string' 
        },
        {
            "name" : "embedding",
            "type" : "float[]",
            "embed": {
                "from": [ "iso", "queueid", "county", "state", "gentype", "description"],
                "model_config": {
                    "model_name": "ts/e5-small"
                }
            }
        }
    ]
}


# Define the Typesense collection schema
update_schema = {
    "fields": [
        { "name": 'queueid', type: 'string'}
    ]
}

def updateCollection(name):
    typesenseClient.collections[name].update(update_schema)
    print('Collection schema updated successfully!')

def dropCollection(name):
    typesenseClient.collections[name].delete()
    print('Collection deleted successfully!')

def syncData(name):
    try:
        #Create or update the collection
        #Create or update the collection
        already_exists = False
        try:
            typesenseClient.collections.create(projects_schema)
        except typesense.exceptions.ObjectAlreadyExists as err:
            print("Collection already exists. Continuing to sync data")
            already_exists = True
        if not already_exists:
            print("Successfully created the collection")
        print("Successfully created the collection")
        # Fetch data from PostgreSQL
        q = "SELECT IsoID, QueueId, County, StateName, GenerationType, AdditionalInfo FROM QueueInfo"
        conn = get_db_connection()
        print("Successfully connected to database")
        cursor = conn.cursor()
        print("Got the cursor")
        cursor.execute(q)
        print("Successfully executed the query")
        results = cursor.fetchall()
        print("Successfully fetched the results")
        docs = typesenseClient.collections[name].documents
        for (isoid, queueid, county, state_name, gentype, desc) in results:
            #print(isoid, ", ", queueid, ", ", county, ", ", state_name, ", ", gentype, ", ", desc)
            #break
            docs.upsert({
                "iso": str(isoid), #Typesense requires string IDs
                "queueid": str(queueid),
                "county": str(county),
                "state": str(state_name),
                "gentype": str(gentype),
                "description": str(desc)
            })
        # Import data into Typesense
            #typesenseClient.collections[name].documents.upsert(docs) 
        print('Data synced successfully!')
    except Exception as e:
        print('Error syncing data:', e)
#    finally:
#        await conn.close()

def import_project_data():
    try:
        collection_name = projects_schema["name"]
        #Create or update the collection
        already_exists = False
        try:
            typesenseClient.collections.create(projects_schema)
        except typesense.exceptions.ObjectAlreadyExists as err:
            print("Collection already exists. Continuing to sync data")
            already_exists = True
        if not already_exists:
            print("Successfully created the collection")
        
        # Fetch data from PostgreSQL
        q = "SELECT IsoID, QueueId, County, StateName, GenerationType, AdditionalInfo FROM QueueInfo"
        conn = get_db_connection()
        print("Successfully connected to database")
        cursor = conn.cursor()
        print("Got the cursor")
        cursor.execute(q)
        print("Successfully executed the query")
        results = cursor.fetchall()
        print("Successfully fetched the results")
        docs = typesenseClient.collections[collection_name].documents
        for (isoid, queueid, county, state_name, gentype, desc) in results:
            #print(isoid, ", ", queueid, ", ", county, ", ", state_name, ", ", gentype, ", ", desc)
            #break
            docs.upsert({
                "iso": str(isoid), #Typesense requires string IDs
                "queueid": str(queueid),
                "county": str(county),
                "state": str(state_name),
                "gentype": str(gentype),
                "description": str(desc)
            })
        # Import data into Typesense
            #typesenseClient.collections[name].documents.upsert(docs) 
        print('Data synced successfully!')
    except Exception as e:
        print('Error syncing data:', e)
#    finally:
#        await conn.close()

def setup_typesense():
    try:
        dropCollection(projects_schema["name"])
    except typesense.exceptions.ObjectNotFound as err:
        print("Collection does not exist. Continuing to import data")
    import_project_data()

if __name__ == "__main__":
    print("Import this module to invoke the method")

File: backend/scripts/sync-typesense.js
#!/usr/bin/env node

/**
 * PowerNOVA Typesense Sync Script
 * 
 * This script syncs project data from PostgreSQL QueueInfo table to Typesense
 * for fast search functionality.
 * 
 * Usage:
 *   node scripts/sync-typesense.js [command]
 * 
 * Commands:
 *   setup     - Drop existing collection and create new one with data
 *   sync      - Sync data to existing collection (create if doesn't exist)
 *   drop      - Drop the collection
 *   status    - Check Typesense connection and collection status
 * 
 * Environment Variables:
 *   TYPESENSE_API_KEY - Typesense API key
 *   TYPESENSE_HOST - Typesense host (default: localhost)
 *   TYPESENSE_PORT - Typesense port (default: 8108)
 *   TYPESENSE_PROTOCOL - Protocol (default: http)
 *   POWERNOVA_* - Database connection parameters
 */

const Typesense = require('typesense');
const { query } = require('../src/config/database');
const path = require('path');
const dotenvPath = path.join(__dirname, '../.env');
console.log('🔍 Attempting to load environment variables from:', dotenvPath);
const dotenvResult = require('dotenv').config({ path: dotenvPath });
if (dotenvResult.error) {
  console.error('❌ Failed to load .env file:', dotenvResult.error);
} else {
  console.log('✅ .env file loaded successfully');
}

// Typesense client configuration
const typesenseConfig = {
  apiKey: process.env.TYPESENSE_API_KEY || 'powernova-api-key',
  nodes: [{
    host: process.env.TYPESENSE_HOST || 'localhost',
    port: process.env.TYPESENSE_PORT || '8108',
    protocol: process.env.TYPESENSE_PROTOCOL || 'http'
  }],
  connectionTimeoutSeconds: 5
};

// Don't create client at module level - create it in functions when needed
// const client = new Typesense.Client(typesenseConfig);

// Collection schema
const projectsSchema = {
  name: 'projects',
  fields: [
    { name: 'iso', type: 'string', facet: true },
    { name: 'queueid', type: 'string' },
    { name: 'project_name', type: 'string', optional: true },
    { name: 'interconnecting_entity', type: 'string', optional: true },
    { name: 'county', type: 'string', facet: true, optional: true },
    { name: 'state', type: 'string', facet: true, optional: true },
    { name: 'gentype', type: 'string', facet: true, optional: true },
    { name: 'capacity_mw', type: 'float', optional: true },
    { name: 'status', type: 'string', facet: true, optional: true },
    { name: 'description', type: 'string', optional: true },
    { name: 'queue_date', type: 'string', optional: true },
    { name: 'proposed_completion_date', type: 'string', optional: true }
  ]
};

async function checkTypesenseConnection() {
  try {
    const client = new Typesense.Client(typesenseConfig);
    const health = await client.health.retrieve();
    console.log('✅ Typesense connection successful');
    console.log(`📊 Typesense status: OK`);
    return true;
  } catch (error) {
    console.error('❌ Typesense connection failed:', error.message);
    console.error('💡 Make sure Typesense is running: docker-compose up typesense');
    return false;
  }
}

async function dropCollection() {
  try {
    const client = new Typesense.Client(typesenseConfig);
    await client.collections('projects').delete();
    console.log('🗑️  Collection "projects" dropped successfully');
    return true;
  } catch (error) {
    if (error.httpStatus === 404) {
      console.log('ℹ️  Collection "projects" does not exist');
      return true;
    }
    console.error('❌ Failed to drop collection:', error.message);
    return false;
  }
}

async function createCollection() {
  try {
    const client = new Typesense.Client(typesenseConfig);
    await client.collections().create(projectsSchema);
    console.log('✅ Collection "projects" created successfully');
    return true;
  } catch (error) {
    if (error.httpStatus === 409) {
      console.log('ℹ️  Collection "projects" already exists');
      return true;
    }
    console.error('❌ Failed to create collection:', error.message);
    return false;
  }
}

async function syncProjectData() {
  try {
    const client = new Typesense.Client(typesenseConfig);
    console.log('📡 Fetching project data from PostgreSQL...');
    
    const sql = `
      SELECT 
        IsoID, QueueID, ProjectName, InterconnectingEntity, 
        County, StateName, GenerationType, CapacityMW, Status,
        AdditionalInfo, QueueDate, ProposedCompletionDate
      FROM QueueInfo 
      ORDER BY IsoID, QueueID
    `;
    
    const result = await query(sql);
    const projects = result.rows;
    
    console.log(`📊 Found ${projects.length} projects to sync`);
    
    if (projects.length === 0) {
      console.log('⚠️  No projects found in database. Run queue population first:');
      console.log('   npm run populate-queue-data populate CAISO');
      return false;
    }

    // Convert database rows to Typesense documents
    const documents = projects.map((project, index) => {
      // Create searchable description from available fields
      const descriptionParts = [
        project.projectname,
        project.interconnectingentity,
        project.generationtype,
        project.county,
        project.statename
      ].filter(Boolean);
      
      return {
        id: `${project.isoid}-${project.queueid}`,
        iso: project.isoid || '',
        queueid: project.queueid || '',
        project_name: project.projectname || '',
        interconnecting_entity: project.interconnectingentity || '',
        county: project.county || '',
        state: project.statename || '',
        gentype: project.generationtype || '',
        capacity_mw: parseFloat(project.capacitymw) || 0,
        status: project.status || '',
        description: descriptionParts.join(' '),
        queue_date: project.queuedate ? project.queuedate.toISOString().split('T')[0] : '',
        proposed_completion_date: project.proposedcompletiondate ? 
          project.proposedcompletiondate.toISOString().split('T')[0] : ''
      };
    });

    console.log('💾 Importing documents to Typesense...');
    
    // Import in batches for better performance
    const batchSize = 100;
    let imported = 0;
    let failed = 0;
    
    for (let i = 0; i < documents.length; i += batchSize) {
      const batch = documents.slice(i, i + batchSize);
      
      try {
        const results = await client.collections('projects').documents().import(batch, {
          action: 'upsert'
        });
        
        // Count successful imports
        const batchImported = results.filter(r => r.success === true).length;
        const batchFailed = results.filter(r => r.success === false).length;
        
        imported += batchImported;
        failed += batchFailed;
        
        const progress = Math.round(((i + batch.length) / documents.length) * 100);
        console.log(`📈 Progress: ${progress}% (${imported} imported, ${failed} failed)`);
        
      } catch (error) {
        console.error(`❌ Batch import failed:`, error.message);
        failed += batch.length;
      }
    }
    
    console.log(`\n🎉 Sync completed!`);
    console.log(`   ✅ Successfully imported: ${imported} documents`);
    if (failed > 0) {
      console.log(`   ❌ Failed to import: ${failed} documents`);
    }
    
    return true;
    
  } catch (error) {
    console.error('❌ Failed to sync project data:', error.message);
    return false;
  }
}

async function getCollectionStatus() {
  try {
    const client = new Typesense.Client(typesenseConfig);
    const collection = await client.collections('projects').retrieve();
    console.log('📊 Collection Status:');
    console.log(`   Name: ${collection.name}`);
    console.log(`   Documents: ${collection.num_documents}`);
    console.log(`   Fields: ${collection.fields.length}`);
    
    // Show some example documents
    const docs = await client.collections('projects').documents().search({
      q: '*',
      per_page: 3
    });
    
    console.log(`\n📄 Sample documents (${docs.hits.length} shown):`);
    docs.hits.forEach((hit, index) => {
      const doc = hit.document;
      console.log(`   ${index + 1}. ${doc.iso}-${doc.queueid}: ${doc.project_name || 'Unnamed'} (${doc.gentype || 'Unknown type'})`);
    });
    
    return true;
  } catch (error) {
    if (error.httpStatus === 404) {
      console.log('ℹ️  Collection "projects" does not exist');
      return false;
    }
    console.error('❌ Failed to get collection status:', error.message);
    return false;
  }
}

async function main() {
  const command = process.argv[2] || 'sync';
  
  console.log(`🚀 PowerNOVA Typesense Sync Tool`);
  console.log(`Command: ${command}\n`);
  
  // Check Typesense connection first
  const connected = await checkTypesenseConnection();
  if (!connected) {
    process.exit(1);
  }
  
  switch (command.toLowerCase()) {
    case 'setup':
      console.log('🔄 Setting up Typesense collection...');
      await dropCollection();
      await createCollection();
      await syncProjectData();
      break;
      
    case 'sync':
      console.log('🔄 Syncing project data...');
      await createCollection(); // Create if doesn't exist
      await syncProjectData();
      break;
      
    case 'drop':
      console.log('🗑️  Dropping collection...');
      await dropCollection();
      break;
      
    case 'status':
      console.log('📊 Checking collection status...');
      await getCollectionStatus();
      break;
      
    default:
      console.log(`❌ Unknown command: ${command}`);
      console.log(`Available commands: setup, sync, drop, status`);
      process.exit(1);
  }
  
  console.log('\n✨ Done!');
}

// Run the script
if (require.main === module) {
  main().catch(error => {
    console.error('💥 Script failed:', error);
    process.exit(1);
  });
}

module.exports = {
  checkTypesenseConnection,
  dropCollection,
  createCollection,
  syncProjectData,
  getCollectionStatus
};
